arroyo-controller/src/job_controller/checkpoint_state.rs

```diff
@@ -0,0 +1,396 @@
+            .flat_map(|metadata| metadata.backend_data.clone())
+            .filter_map(Self::backend_data_to_key)
+            .collect();
+
+        let size = subtasks
+            .values()
+            .fold(0, |size, s| size + s.metadata.as_ref().unwrap().bytes);
+
+        StateBackend::write_operator_checkpoint_metadata(OperatorCheckpointMetadata {
+            job_id: self.job_id.to_string(),
+            operator_id: operator_id.clone(),
+            epoch: self.epoch,
+            start_time: to_micros(start_time),
+            finish_time: to_micros(finish_time),
+            min_watermark,
+            max_watermark,
+            has_state,
+            tables: tables.into_values().collect(),
+            backend_data: backend_data.into_values().collect(),
+            bytes: size,
+        })
+        .await;
+
+        if let Some(op) = self.operator_details.get_mut(&operator_id) {
+            op.finish_time = Some(to_micros(finish_time));
+        }
+
+        self.completed_operators.insert(operator_id);
+    }
+
+    fn backend_data_to_key(backend_data: BackendData) -> Option<((u32, String), BackendData)> {
+        let Some(internal_data) = &backend_data.backend_data else {
+            return None;
+        };
+        match &internal_data {
+            backend_data::BackendData::ParquetStore(data) => {
+                Some(((data.epoch, data.file.clone()), backend_data))
+            }
+        }
+    }
+
+    pub fn done(&self) -> bool {
+        self.completed_operators.len() == self.tasks_per_operator.len()
+    }
+
+    pub fn committing_state(&self) -> CommittingState {
+        CommittingState::new(self.checkpoint_id, self.subtasks_to_commit.clone())
+    }
+
+    /// Syncs the operator details to the database so the API/UI can see them
+    pub async fn update_db(&self, pool: &Pool) -> anyhow::Result<()> {
+        let c = pool.get().await?;
+
+        controller_queries::update_checkpoint()
+            .bind(
+                &c,
+                &serde_json::to_value(&self.operator_details).unwrap(),
+                &None,
+                &crate::types::public::CheckpointState::inprogress,
+                &self.checkpoint_id,
+            )
+            .await?;
+
+        Ok(())
+    }
+
+    // TODO: maybe pool should be a trait so we can mock it for testing?
+    pub async fn finish(self, pool: Option<&Pool>) -> anyhow::Result<()> {
+        let finish_time = SystemTime::now();
+        StateBackend::complete_checkpoint(CheckpointMetadata {
+            job_id: self.job_id,
+            epoch: self.epoch,
+            start_time: to_micros(self.start_time),
+            finish_time: to_micros(finish_time),
+            min_epoch: self.min_epoch,
+            operator_ids: self.completed_operators.iter().cloned().collect(),
+        })
+        .await;
+
+        let operator_state = serde_json::to_value(&self.operator_details).unwrap();
+
+        if let Some(pool) = pool {
```

Comment:
This seems a bit of an anti-pattern. This is now two methods "finish saving operator state" and "update DB with new checkpoint status" and depending on how it is called you only do the second sometimes. Can you try factoring it out into two methods?

Reply:
Good idea, I've split this into two functions and changed `finish_checkpoint_if_done` to use them.